Support Server Name Indication & fix handshake

jetm opened the pull request:
While adding `https://` support to U-Boot, several issues were found while using http client app from lwip with SSL support from mbedtls. Further details are in the commit messages.

Commit:
altcp_tls_mbedtls: Support Server Name Indication

SNI, or Server Name Indication, is an addition to the TLS encryption
protocol that enables a client device to specify the domain name it is
trying to reach in the first step of the TLS handshake, preventing
common name mismatch errors and not reaching to HTTPS server that
enforce this condition.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

goldsimon commented on src/apps/altcp_tls/altcp_tls_mbedtls.c:
```diff
@@ -974,13 +976,15 @@ altcp_tls_create_config_client_common(const u8_t *ca, size_t ca_len, int is_2way
     mbedtls_ssl_conf_ca_chain(&conf->conf, conf->ca, NULL);
   }
+
+  memcpy(conf->host, host, sizeof(conf->host));
```
Shouldn't this be range-checked and NULL-checked?

jetm replied:
Done. Thanks

goldsimon commented on src/core/tcp_out.c:
```diff
@@ -1255,14 +1255,6 @@ tcp_output(struct tcp_pcb *pcb)
   LWIP_ASSERT("don't call tcp_output for listen-pcbs",
               pcb->state != LISTEN);
 
-  /* First, check if we are invoked by the TCP input processing
```
Removing this part won't work. It's been in here for too long to just remove it like that. There may be drivers or ports out there that depend on this behaviour (output not being called while in input processing). Let's take a step back and see what you are trying to fix here!

jetm replied:
No apply anymore

jetm replied:
Remove that part. It was for testing. Thanks

goldsimon commented on src/apps/altcp_tls/altcp_tls_mbedtls.c:
```diff
@@ -3,8 +3,8 @@
- * 
- * This version is currently compatible with the 2.x.x branch (current LTS).
+ *
+ * This version is currently compatible with the 3.6.x branch (latest LTS).
```
The part about only TLS 1.2. with LWIP http client app and OS_SYS=0 (U-Boot) should probably be marked here. Being like that, I'd rather merge patch #10465 from savannah (https://savannah.nongnu.org/patch/?10465), that would probably fix it for you as well?

jetm replied:
Not apply anymore

jetm replied:
Yes, that patch is better. Thank you for bringing it out. I have removed my patch.

goldsimon reviewed (changes requested):
Hi Javier,
please see the 3 review comments.
Regards, Simon

Commit:
altcp_tls_mbedtls: Fix handshake when called from u-boot

When using the http-client LWIP app in U-Boot (OS_SYS=0), the handshake
fails because LWIP doesn't send TCP packets after it initiates.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

jetm commented:
> Hi Javier, please see the 3 review comments. Regards, Simon

Hi @goldsimon I have addressed your three comments. Please, when you have time, take a look at the new changes. Thanks.
